KMAC submission

hoxxep opened the pull request:
I have tried to follow the RustCrypto MACs workflow, and used the traits as best I can.

This implements KMAC128, KMAC256, KMACXOF128, and KMACXOF256.

Key deviations from the existing mac traits:
- `Mac::finalize()` outputs a 32-bit digest for KMAC128, and 64-bit for KMAC256 following the NIST default MAC output length guidance in section 8.4.
- `Kmac*::finalize_into_buf(out: &mut [u8])` is a generic finalize method for any fixed output size, following section 4.3 of [NIST SP 800-185]. I wasn't aware of a suitable trait to express this.
  - There is `FixedOutput::finalize_into`, but it's limited to the Mac-defined output size `Output<Self>`. This implementation allows for runtime determined output sizes.
- `ExtendableOutput::finalize_xof()` is used to implement KMACXOF128 and KMACXOF256. We could add an ExtendableMac trait to incorporate ExtendableOutput?
- `Kmac*::new_customization(secret: &[u8], customization: &[u8])` is used to initialize with a customization string, and again I wasn't aware of any suitable existing trait to express this.

I'd be happy to try to add traits for this, if you believe this is the right direction? The crate should already be useful in it's current form though, and we could add traits later.

An alternate design could separate Kmac128, Kmac256, KmacXof128, and KmacXof256 types. This PR combines the fixed and XOF outputs into the same type, with different finalization methods for each output.

Happy to hear any feedback!

Related issues: 
- https://github.com/RustCrypto/MACs/issues/133
- https://github.com/RustCrypto/MACs/issues/1

[NIST SP 800-185]: https://nvlpubs.nist.gov/nistpubs/SpecialPublications/NIST.SP.800-185.pdf

hoxxep committed:
kmac: Initial crate submission

hoxxep committed:
kmac: Replace sha3 dependency with new cshake crate

newpavlov commented on kmac/benches/mod.rs:
```diff
@@ -0,0 +1,61 @@
+use test::Bencher;
+
+#[macro_export]
+macro_rules! bench_full {
```
Is there a reason why `digest::bench_update!` can not be used here?

hoxxep committed:
kmac: Refactor Rate trait based implementation

Integrating comments from @newpavlov's code review.

- Remove the messy impl kmac macro in favour of a cleaner trait system
  approach.
- Clean up test vectors and use hex-literal, and use raw NIST vectors
  copy/pasted from the PDFs to make review easier.
- Move public API tests into tests/kmac.rs
- Remove std from the std tests.
- Remove the bench_full benchmarks, we will discuss this in a separate PR.
- Adjust docstrings and README for completeness.

hoxxep commented on kmac/src/lib.rs:
```diff
@@ -0,0 +1,237 @@
+impl<Rate: BlockSizes + sealed::KmacParams> MacMarker for Kmac<Rate> {}
+
+impl<Rate: BlockSizes + sealed::KmacParams> OutputSizeUser for Kmac<Rate> {
+    type OutputSize = <Rate as sealed::KmacParams>::OutputSize;
```
Alternatively, I had considered allowing the `OutputSize` to be fully generic*, as it's valid to generate any output length for KSF purposes.

I still want to ensure the `Kmac128` and `Kmac256` types provide the recommended default MAC output sizes, but I also don't want the average user to need to understand the `Rate`. One option is a second set of type aliases that only expose a generic `OutputSize` (`Kmac128Something<OutputSize>`). 

Alternatively the core Kmac parameters could be the `Capacity` and `OutputSize` which are easier for a user to understand (eg. `Kmac<U128, OutputSize>`). ~The `Rate` can be computed via `type Rate = Diff<U200, Capacity>`.~ (edit: brain fart sorry, it needs something more complicated for the bits/bytes conversion). I slightly prefer this approach from a user's perspective, but I don't know if using a `Capacity` parameter (instead of `Rate`) would break tradition from the other RustCrypto crates?

*I would still also keep `finalize_into_buf` for its runtime-determined output size.

hoxxep committed:
kmac: Use the new chake const-generic API

hoxxep commented:
@newpavlov I've ported kmac to the latest cshake const-generic based API, thanks!

hoxxep committed:
kmac: use published cshake crate
